File: integration/testdata/kotlin/externalcalls/Echo.kt
package ftl.echo

import ftl.echo2.echo as echo2
import xyz.block.ftl.Context
import xyz.block.ftl.Verb

data class EchoRequest(val name: String)
data class EchoResponse(val message: String)

@Verb
fun echo(context: Context, req: EchoRequest): EchoResponse {
  return EchoResponse(message = "Hello, ${req.name}!")
}

@Verb
fun call(context: Context, req: EchoRequest): EchoResponse {
  val res = context.call(::echo2, ftl.echo2.EchoRequest(name = req.name))
  return EchoResponse(message = res.message)
}

File: examples/go/echo/echo.go
// This is the echo module.
package echo

import (
	"context"
	"fmt"

	"ftl/time"

	"github.com/block/ftl/go-runtime/ftl"
)

type Default = ftl.Config[string]

// An echo request.
type EchoRequest struct {
	Name ftl.Option[string] `json:"name"`
}

type EchoResponse struct {
	Message string `json:"message"`
}

// Echo returns a greeting with the current time.
//
//ftl:verb export
func Echo(ctx context.Context, req EchoRequest, tc time.TimeClient, defaultName Default) (EchoResponse, error) {
	tresp, err := tc(ctx, time.TimeRequest{})
	if err != nil {
		return EchoResponse{}, err
	}

	return EchoResponse{Message: fmt.Sprintf("Hello, %s!!! It is %s!", req.Name.Default(defaultName.Get(ctx)), tresp.Time)}, nil
}
